README
CDEmu Daemon
v.1.2.0
~~~~~~~~~~~~~


Table of contents
~~~~~~~~~~~~~~~~~

1. Introduction
2. Userspace-cdemu suite overview
3. Requirements
4. Installation and set-up
5. Troubleshooting
6. CDEmu daemon D-BUS interface
7. Device options
8. Debugging
9. Encrypted images and password supplying
10. Contact information


1. Introduction
~~~~~~~~~~~~~~~

This is CDEmu daemon, the userspace daemon part of the userspace-cdemu suite, a 
free, GPL CD/DVD-ROM device emulator for linux.

This document describes both the userspace-cdemu suit in general, and the 
specifics of CDEmu daemon, such as installation quirks, troubleshooting, and 
daemon's D-Bus interface.


2. Userspace-cdemu suite overview
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The whole userspace-cdemu suite is intended as a rewrite of cdemu kernel module
and userspace utility that has been written by Robert Penz. But instead of 
providing only block device and linux uniform CD-ROM driver interface, CDEmu 
daemon tries to implement packet command set as specified by MMC-3 and INF-8090,
thus fully emulating an optical storage device.

Main features:
 - full SCSI device emulation: CDEmu daemon attempts to implement MMC-3 packet
     command set, therefore allowing virtual devices to be used not only as 
     block devices, but as full SCSI devices. Aside from mounting the device,
     you can use dd on it, use it with CD-player, or even copy the loaded image
     using one of linux's CD-ROM burning utilities.
 - multiple image format support: userspace-cdemu is based libMirage, an image
     access library, which attempts to provide unified access to various CD-ROM
     image formats (including B6T, CCD, CDI, CUE, ISO, MDS, NRG and TOC)
 - linux HAL support: userspace-cdemu devices are properly detected by HAL, thus
     giving you the same benefits as with the real devices; most notable are
     automount and detection in WINE
 - debugging: both CDEmu daemon and underlying libMirage library have debugging
     code that supports changing verbosity of debug messages while daemon is 
     running, making it easier to discover and fix problems.

The whole userspace-cdemu suite consists of three major components:
 - kernel module
 - userspace daemon
 - userspace clients
 
2.1. Kernel module
~~~~~~~~~~~~~~~~~~

Kernel module takes care of device emulation in the kernel; it registers virtual
device with appripriate drivers and creates corresponding device nodes. It also 
creates a special character device that is used for communication with userspace. 

As device is accessed, requests are generated by kernel, which are in fact SCSI 
commands. These are passed to userspace daemon via afore-mentioned character 
device. Once daemon processes the request, it returns appropriate data and 
status to kernel, thus completing the request.

The whole process is very similar to accessing real device, except that requests
are passed to userspace daemon instead to hardware.

Early, experimental versions of userspace-cdemu used a module that was called
cdemu-module. This release is based on vhba module, which was written by Chia-I 
Wu. Contrary to cdemu-module, which implemented all the interfaces (i.e. block 
device, uniform CD-ROM driver, etc.) manually, vhba implements virtual SCSI host
and lets the kernel's SCSI layer do the rest. This approach is cleaner, faster
and more robust.

2.2. Daemon
~~~~~~~~~~~

This is the part that these docs (should) belong to. The daemon receives SCSI
commands from kernel module and processes them, passing the requested data back
to the kernel. 

Daemon implements the actual virtual device; one instance per each device 
registered by kernel module. It uses libMirage, an image access library that is
part of userspace-cdemu suite, for the image access (e.g. sector reading).

Daemon is controlled through methods that are exposed via D-BUS. It is written
in C and based on GLib (and thus GObjects), but being controlled over D-BUS, it
allows for different clients written in different languages.

2.3. Clients
~~~~~~~~~~~~

Clients are used to control the daemon; examples of the tasks that can be 
achieved by using daemon's D-BUS interface are: loading and unloading a device,
checking device(s) status, setting the debug mask, etc.

A client can be written in any language, as long as it has D-BUS bindings; it can 
be either a CLI or a GUI application. For illustration, two clients are included
as part of userspace-cdemu:
 - cdemu-client: a simple CLI client
 - gcdemu: a GNOME applet
For more information about these particular clients, consult their documentation.

Should you be interested in writing your own client, there is a section in this
document that describes the daemon's interface, methods and signals.


3. Requirements
~~~~~~~~~~~~~~~

CDEmu daemon has the following software requirements:
 - GLib 2.6 or newer
 - D-BUS 0.60 or newer
 - libdaemon 0.10 or newer
 - libao 0.8.0 or newer
 - libsysfs 
 
 - libMirage 1.2.0 or newer (part of the suite)
 - VHBA module (part of the suite)

 
4. Installation and set up
~~~~~~~~~~~~~~~~~~~~~~~~~~

The build and installation process is classical ./configure, make, make install.
However, there are some quirks one should be aware of.


Make sure you have installed all the required libraries; GLib, D-BUS, libao and 
libdaemon should be provided by most of the modern distributions and should as
such be detected properly by configure script. 

libMirage, which is also part of userspace-cdemu suite, is also required. If you 
are building and installing it yourself, make sure that its .pc file is visible 
to pkg-config utility; if you are using /usr prefix, you should be fine, but in 
case /usr/local or something else is used and you haven't set up pkg-config to 
look there, you might need to force it to do so.

For example, if libMirage was installed to /usr/local, configure would have to 
be run as "PKG_CONFIG_PATH=/usr/local/lib/pkg-config ./configure".

Vhba module is not a build-time dependency, but it might be a good idea to have 
it installed at this point. 


The daemon can register its interface on either system or session D-BUS' bus. By 
default, system bus is used. To use session bus, use --bus=session when running daemon. 
If daemon is ran in daemon mode, usage of system bus is automatically implied.

If you wish to run CDEmu daemon on system bus, CDEmu daemon's configuration files
for D-BUS must be installed in appropriate directory. On most systems, this is
/etc/dbus-1/system.d. However, due to the way autotools are used, /etc gets 
prefixed and therefore config file ends up installed into ${prefix}/etc/. To 
avoid this, you should pass --sysconfdir=/etc to configure, regardless of the 
prefix you are passing to configure. Alternatively, you can specify appropriate 
directory using --with-dbus-system-dir switch.

We use a single libao backend to support CD-audio playback. This allows you to 
select from a number of available audio drivers. Valid choices are the driver
names used by libao as well as "default" and "null". CDEmu is usually 
configured to use the "default" libao audio driver, which means that the 
audio driver specified in libao's config file ("/etc/libao.conf" or "~/.libao") 
is used. If "null" is specified, audio is sent to the null driver, essentially
turning off audio playback. For a full list of valid driver names have a look 
at the libao documentation, in particular the "libao.conf" manpage. 

Also note that most media players on Linux are capable of digital audio 
extraction, meaning they can play CD-audio directly without relying on the 
"analog" playback capabilities of the virtual optical device. Because of this
it's possible to use the libao "null" driver and still play CDs normally.

Unlike previous (experimental snapshot) releases, this release of CDEmu daemon
does not provide initscripts. After some thought it was decided it was too much
hassle to write initscripts for different distributions, especially with alternatives
to SysV init system becoming more and more popular. Packagers are welcome to add 
initscripts specific to distribution they are packaging for, of course. Bear in
mind, though, that daemon started by initscripts should be ran in daemon mode
(and therefore use system bus).


5. Troubleshooting (When bad things happen to good people)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Q: I built and installed the module, but I can't insert it ('modprobe vhba' 
results in something along the lines of 'FATAL: Module vhba not found.').

A: Try running 'depmod -a' after you have installed the module.


Q: Daemon fails to start with the following message:
    'cdemud: cdemud_daemon_initialize: failed to open control device /dev/vhba_ctl!
     Daemon initialization failed: Failed to open control device.'

A: The message indicates that either vhba module is not inserted or you don't 
   have read/write permissions for /dev/vhba_ctl. Please make sure that module 
   is inserted, that /dev/vhba_ctl exists and that you have proper read/write 
   permissions.
   
   
Q: Daemon fails to start with the following message:
    'cdemud: cdemud_daemon_initialize: failed to get name on system bus!
     Daemon initialization failed: Name request on D-BUS failed.'

A: The message indicates that name registration on the system bus has failed. 
   The most likely cause of this is improper permissions. Make sure that CDEmu 
   daemon's D-BUS configuration file is in proper place and that you are running 
   the daemon as root.


Q: Daemon fails to start with the following message:
   cdemu0: cdemud_audio_initialize: failed to open audio device 
   (driver: 'default')!

A: Most likely you will need to specify a working audio driver in libao's config
   file (/etc/libao.conf or ~/.libao), or, if you are using initscripts that came
   with the package for your distribution, in the config file it provided. 
   
   Also, please note that running system-wide (daemonized) instance of the 
   daemon and pulseaudio output doesn't work if pulseaudio is run per-session.


6. CDEmu daemon D-BUS interface (a.k.a. writing your own client)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Communication between CDEmu daemon and a client is done via D-BUS. CDEmu daemon
registers itself on either sytem or session bus, where it exposes the interface 
which can be used by client to control the daemon.

6.1. D-BUS name and object path
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The name CDEmu daemon requests on the system bus is "net.sf.cdemu.CDEMUD_Daemon",
whereas the object path is "/CDEMUD_Daemon". The name of the interface is 
"net.sf.cdemu.CDEMUD_Daemon".

To illustrate, a simple python program connecting to daemon registered on system
bus would therefore look like this:

import dbus
if getattr(dbus, 'version', (0,0,0)) >= (0,41,0):
    import dbus.glib

system_bus = dbus.SystemBus()
dbus_proxy = system_bus.get_object('net.sf.cdemu.CDEMUD_Daemon', '/CDEMUD_Daemon')
dbus_iface = dbus.Interface(dbus_proxy, 'net.sf.cdemu.CDEMUD_Daemon')

6.2. Methods
~~~~~~~~~~~~

For more information on D-BUS methods, argument direction and signatures or 
language-specific data type mappings, please consult the D-BUS documentation.

CDEmu daemon's interface consists of the following methods that can be used to
control the daemon:

* GetDaemonVersion (version)
    + version: out; "s"
        Daemon version string.
    
    - Returns daemon version string.

* GetLibraryVersion (version)
    + version: out; "s"
        Library version string
    
    - Returns library version string of libMirage library used by
      the daemon.

* GetDaemonInterfaceVersion (version)
    + version: out; "i"
        Daemon interface version number
    
    - Returns daemon interface version number, which should be used to detect
      if client can use the interface or not.

* EnumDaemonDebugMasks (type, masks)
    + masks: out; "a(si)"
        Array of structures containing supported debug masks. Each structure
        has two fields; debug mask name (string) and debug mask value (int).
        
    - Returns debug masks supported by the daemon.

* EnumLibraryDebugMasks (type, masks)
    + masks: out; "a(si)"
        Array of structures containing supported debug masks. Each structure
        has two fields; debug mask name (string) and debug mask value (int).
        
    - Returns debug masks supported by the libMirage library that is used by 
      the daemon.
  
* EnumSupportedParsers (parsers)
    + parsers: out; "a(ssss)"
        Array of structures containing information about supported parsers.
        Each structure has multiple fields; parser ID (string), parser name 
        (string), image type description (string) and image file MIME type 
        (string).
        
    - Returns supported image parsers.
    
* EnumSupportedFragments (fragments)
    + framgents: out; "a(ss)"
        Array of structures containing information about supported fragments.
        Each structure has two fields; fragment ID (string) and fragment name 
        (string).
        
    - Returns supported fragments.
    
* GetNumberOfDevices (number_of_devices)
    + number_of_devices: out; "i"
        Number of devices (int).
    
    - Returns number of virtual devices.
  
* DeviceGetStatus (device_number, loaded, file_names)
    + device_number: in; "i"
        Device you are requesting status for (int).
    + loaded: out; "b"
        Boolean denoting whether device is loaded or not.
    + file_names: out; "as"
        Image's file name(s) (string).
    
    - Returns the status of specified device.
  
* DeviceLoad (device_number, file_names, parameters)
    + device_number: in; "i"
        Device that is to be loaded (int).
    + file_names: in; "as"
        File name(s) of the image to be loaded (array of strings).
    + parameters: in; "a{sv}"
        Additional (optional) parameters, stored in a D-BUS dictionary type. If 
        no parameters are needed, empty dictionary should be sent.
        Currently supported parameters:
            - "password":"password_string"; password for encrypted images
        
    - Attempts to load the image into specified device. 
    - The client might wish to detect MIRAGE_E_NEEDPASSWORD error, which 
      indicates that the image is encrypted and needs a password provided via 
      parameters.
    
* DeviceUnload (device_number)
    + device_number: in; "i"
        Device that is to be unloaded (int).
        
    - Attempts to unload the device.
  
* DeviceGetOption (device_number, option_name, option_values)
    + device_number: in; "i"
        Device for which the specified option's value is to be retrieved (int).
    + option_name: in; "s"
        Name of the option (string).
    + option_values: out; "av"
        Option value(s) (array of variants). The actual content depends on the 
        option.
    
    - This method retrieves the value(s) of specified option for specified device.
      For more information on supported options, see Section 7.
  
* DeviceSetOption (device_number, option_name, option_values)
    + device_number: in; "i"
        Device for which the specified option's value is to be set (int).
    + option_name: in; "s"
        Name of the option (string).
    + option_values: in; "av"
        Option value(s) (array of variants). The actual content depends on the 
        option.
    
    - This method sets the value(s) of specified option for specified device.
      For more information on supported options, see Section 7.

6.3. Signals
~~~~~~~~~~~~

In addition to the methods, CDEmu daemon's interface emits the following signals 
in response to the change of status of either the daemon or one of the devices:

* DaemonStarted
    - emitted when daemon is started

* DaemonStopped
    - emitted when daemon is stopped

* DeviceStatusChange
    + device_number: "i"
        Device that emitted the signal (int).
        
    - emitted when device is loaded/unloaded

* DeviceOptionChange
    + device_number: "i"
        Device that emitted the signal (int).
    + option: "s"
        Option that has been changed (string).
        
    - emitted when device's option is changed

 
7. Device options
~~~~~~~~~~~~~~~~~

CDEmu virtual devices have the following options that can be get/set via the
DeviceGetOption/DeviceSetOption API. Please note that number of arguments and
argument types for DeviceSetOption function are strictly checked; i.e. tbe
method call will fail, for example, if argument for the 'dpm-emulation option' 
is of type G_TYPE_INT.

* dpm-emulation
   + arguments: enabled (G_TYPE_BOOLEAN)
   
   - DPM emulation flag. Determines whether DPM emulation is performed.
   
* tr-emulation
   + arguments: enabled (G_TYPE_BOOLEAN)
   
   - Transfer rate emulation flag. Determines whether transfer rate emulation is
     performed
     
* device-id
   + arguments: vendor_id (G_TYPE_STRING), product_id (G_TYPE_STRING), revision
     (G_TYPE_STRING), vendor_specific (G_TYPE_STRING)
     
   - Device ID. Determines the device identifier (vendor identifier, product 
     identifier, product revision and vendor-specific string) as returned by 
     INQUIRY command. Changing the ID might be useful in cases when an 
     application doesn't like the default ID for some reason.
     
   - Note that the length of argument strings should be as following: vendor_id 
     8 characters, product_id 16 characters, revision 4 characters and 
     vendor_specific 20 characters. The actual arguments can surpass these 
     lengths, but only specified amount of characters will actually be used by
     the device.
     
* daemon-debug-mask
   + arguments: mask (G_TYPE_INT)
   
   - Daemon debug mask. Determines the amount of verbosity of daemon's code.
   
* library-debug-mask
   + arguments: mask (G_TYPE_INT)
   
   - Library debug mask. Determines the amount of verbosity of libMirage library
     used by the daemon.

8. Debugging
~~~~~~~~~~~~


ChangeLog
==========================================================================
 Version 1.2.0   2009-07-26
==========================================================================

Rok Mandeljc:
	* Adaptations to new libmirage parser framework.
	* Password support.
	* Versioning support.
	* Updated Slovenian translation.
	* Updated documentation.
	* Misc. fixes.

Loïc Martin:
	* French translation.

==========================================================================
 Version 1.1.0   2008-06-29
==========================================================================

Marcel Hoppe:
	* German translation.

Rok Mandeljc:
	* Added documentation.
	* Slovenian translation.
	* Added 'any' and 'all' keywords.
	* Many improvements to the code.
	* Added reporting of device mappings.
	* Many bugfixes.

Henrik Stokseth:
	* Norwegian translation.
	* Code improvements and bugfixes.

==========================================================================
 Version 1.0.0   2007-12-24
==========================================================================

Rok Mandeljc:
	* Initially wrote the whole CDEmu client.



src/applet.py
# -*- coding: utf-8 -*-
# gCDEmu: Applet class
# Copyright (C) 2006-2008 Rok Mandeljc
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import sys
import os
import traceback

import gconf
import gobject
import gtk, gtk.gdk
import gnome
import gnome.ui
import gnomeapplet

from gettext import gettext as _

import dbus
from dbus.mainloop.glib import DBusGMainLoop
DBusGMainLoop(set_as_default=True)

try:
    import pynotify
    has_pynotify = True
except:
    print "You don't seem to have pynotify installed; notifications disabled."
    has_pynotify = False

# Internals
import globals as config
from device import gCDEmu_Device

class gCDEmu_Applet (gnomeapplet.Applet):    
    def __init__(self):
        self.__gobject_init__()
        return

    def __cb_gconf_client_notify (self, client, id, entry, user_data):        
        if entry.get_key() == self.__gconf_key_path + "/icon_name":
            icon_name = entry.get_value().get_string()
            print "Icon name changed to: %s" % (icon_name)
                        
            try:
                self.__pixbuf_icon = gtk.gdk.pixbuf_new_from_file_at_size(config.image_dir + "/" + icon_name, 204, 204)
            except gobject.GError, e:
                message = gtk.MessageDialog(None, 0, gtk.MESSAGE_ERROR, gtk.BUTTONS_CLOSE, e.message)
                message.set_title(_("Error"))
                message.run()
                message.destroy()
            
            self.__refresh_icon()
            
        elif entry.get_key() == self.__gconf_key_path + "/notifications":
            value = entry.get_value().get_bool()
            print "Show notifications changed to: %i" % (value)
            
            # Does user want us to use notifications?
            self.__show_notifications = value
            
            # Update popup menu
            self.__popup.set_prop("/commands/notifications", "state", "%i" % (self.__show_notifications))
        elif entry.get_key() == self.__gconf_key_path + "/use_system_bus":
            value = entry.get_value().get_bool()
            print "Use system bus changed to: %i" % (value)
            
            # Should we use system bus?
            self.__use_system_bus = value
            
            # Update popup menu
            self.__popup.set_prop("/commands/system_bus", "state", "%i" % (self.__use_system_bus))

            # Switch bus and reconnect
            self.__switch_bus()
            self.__cleanup()
            self.__connect()
        else:
            print "Unhandled key: %s\n" % (entry.get_key())
        
    def __display_error (self, text):
         # Show error dialog
        message = gtk.MessageDialog(None, 0, gtk.MESSAGE_ERROR, gtk.BUTTONS_CLOSE, text)
        message.set_title(_("Error"))
        message.run()
        message.destroy()
        
        return
        
    def setup_applet (self):
        # Initialize variables
        self.__gconf_client = None
        self.__gconf_key_path = None

        self.__show_notifications = True
        self.__use_system_bus = True
        
        self.__about = None
        
        self.__pixbuf_logo = None
        self.__pixbuf_icon = None
        self.__image = None
    
        self.__panel_size = 24
        
        self.__dbus_bus = None
        self.__dbus_iface = None

        self.__connected = False
        self.__signal_started = None
    
        self.__menu = None
        self.__devices = []
        
        # Init applet
        self.set_applet_flags(gnomeapplet.EXPAND_MINOR)
        
        # Init G-conf
        self.add_preferences("/schemas/apps/gcdemu-applet/prefs");
        self.__gconf_client = gconf.client_get_default()
        self.__gconf_key_path = self.get_preferences_key()
        if self.__gconf_key_path == None:
            self.__gconf_client.add_dir('/apps/gcdemu-applet-standalone/prefs', gconf.CLIENT_PRELOAD_NONE)
            self.__gconf_key_path = "/apps/gcdemu-applet-standalone/prefs"
        
        # Watch over our config
        self.__gconf_client.notify_add(self.__gconf_key_path, self.__cb_gconf_client_notify)
        
        # Logo; load the SVG, scaled to 156x156
        self.__pixbuf_logo = gtk.gdk.pixbuf_new_from_file_at_size(config.image_dir + "/" + "gcdemu.svg", 156, 156)
        
        # Icon image
        self.__image = gtk.Image()
        self.add(self.__image)
        self.set_background_widget(self)
        
        # To load the icon, we pretend that icon path g-conf key has changed and
        # let the callback do the rest for us...
        self.__gconf_client.notify(self.__gconf_key_path + "/icon_name")
        
        # Set up About dialog
        self.__about = gtk.AboutDialog()
        self.__about.set_name("gCDEmu")
        self.__about.set_version(config.version)
        self.__about.set_copyright("Copyright (C) 2006-2008 Rok Mandeljc")
        self.__about.set_comments(_("gCDEmu is an applet for controlling CDEmu\ndevices via CDEmu daemon."))
        self.__about.set_website("http://cdemu.sf.net")
        self.__about.set_website_label(_("The CDEmu project website"))
        self.__about.set_authors([ "Rok Mandeljc <rok.mandeljc@gmail.com>" ])
        self.__about.set_documenters([ "Rok Mandeljc <rok.mandeljc@gmail.com>" ])
        self.__about.set_artists([ "Rômulo Fernandes <abra185@gmail.com>" ])
        self.__about.set_translator_credits(_("translator-credits"))
        self.__about.set_logo(self.__pixbuf_logo)
            
        # Set up popup menu
        popup_menu_xml = "<popup name='button3'>"
        popup_menu_xml += "<menuitem name='system_bus' verb='system_bus' label='%s' type='toggle'/>" % (_("Use _system bus"))
        if has_pynotify:
            popup_menu_xml += "<menuitem name='notifications' verb='notifications' label='%s' type='toggle'/>" % (_("Show _notifications"))
        popup_menu_xml += "<separator />"
        popup_menu_xml += "<menuitem name='about' verb='about' label='%s' pixtype='stock' pixname='gnome-stock-about' />" % (_("_About"))
        popup_menu_xml += "<menuitem name='help' verb='help' label='%s' pixtype='stock' pixname='gtk-help' />" % (_("_Help"))
        popup_menu_xml += "</popup>"
        
        # We will install custom UI handlers for other menu elements, so verbs
        # are not needed
        popup_menu_verbs = [
            ( "about", lambda w,d: self.__verb_about() ),
            ( "help",  lambda w,d: self.__verb_help()  ),
        ]
        
        popup = self.get_popup_component()
        popup.connect("ui-event", self.__cb_popup_ui_event)
        self.__popup = popup

        # Setup menu        
        self.setup_menu(popup_menu_xml, popup_menu_verbs, None)
                
        self.show_all()
        
        self.connect("size-allocate", self.__cb_applet_size_allocate) # The best way (TM) to monitor panel size change; "change-size" is useless...
        self.connect("button-press-event", self.__cb_applet_button_press_event)        
        
        # Init PyNotify
        if has_pynotify:            
            if pynotify.init(config.name):
                # Load notifications setting
                self.__show_notifications = self.__gconf_client.get_bool(self.__gconf_key_path + "/notifications")
                print "Show notifications: ", self.__show_notifications
                # Set 'notifications' checkbox on popup menu
                popup.set_prop("/commands/notifications", "state", "%i" % (self.__show_notifications))
            else:
                print "There was a problem initializing the pynotify module."
                self.__show_notifications = False
        else:
            self.__show_notifications = False
        
        # Which bus to use...
        self.__use_system_bus = self.__gconf_client.get_bool(self.__gconf_key_path + "/use_system_bus")
        popup.set_prop("/commands/system_bus", "state", "%i" % (self.__use_system_bus))
                    
        # Try to connect
        print "Trying to connect..."
        self.__switch_bus() # Set up appropriate bus
        self.__connect() # Connect
        
        return True
    

    def __signal_daemon_started (self):
        self.notification(_("Daemon started"), _("CDEmu daemon has been started."))
        # Clean up old connection and start new one
        self.__cleanup()
        self.__connect()
        
        return
        
    def __signal_dbus_daemon_stopped (self):
        self.notification(_("Daemon stopped"), _("CDEmu daemon has been stopped."))
        # Clean up old connection
        self.__cleanup()
        return


    def __cb_applet_button_press_event (self, widget, event):
        if event.button == 1:
            # Button 1: Display menu    
            if self.__menu:
                self.__menu.popup(None, None, self.__calculate_popdown_menu_position, event.button, event.time)
                #self.__menu.popup(None, None, None, event.button, event.time, None)
    
    def __cb_popup_ui_event (self, bonobo_component, verb, event_type, state_string):
        # For some reason, it seems that callback is called every time menu is
        # dropped and one of the toggle buttons is set; we need to do some additional
        # filtering (i.e. compare state string with current flag setting)
        value = bool(int(state_string))
        if verb == "notifications":
            if value != self.__show_notifications:
                print "Notifications flag change"
                # Store the settings
                self.__gconf_client.set_bool(self.__gconf_key_path + "/notifications", value)
        elif verb == "system_bus":
            if value != self.__use_system_bus:
                print "System bus flag change"
                # Store the settings
                self.__gconf_client.set_bool(self.__gconf_key_path + "/use_system_bus", value)
        else:
            print "Unknown verb: %s!" % (verb)
        
        return True
        
    def __cb_applet_size_allocate (self, applet, allocation):
        orient = applet.get_orient()
        
        if orient == gnomeapplet.ORIENT_UP or orient == gnomeapplet.ORIENT_DOWN:
            if applet.allocation.height == self.__panel_size:
                return
            else:
                self.__panel_size = applet.allocation.height
        else:
            if applet.allocation.width == self.__panel_size:
                return
            else:
                self.__panel_size = applet.allocation.width
        
        self.__refresh_icon()

        return
    
    def __verb_about (self):
        # Run About dialog
        self.__about.run()
        self.__about.hide()
        
        return

    def __verb_help (self):
        gnome.help_display(config.name)
        return
    
    def notification (self, summary, body, icon_name=None):
        if self.__show_notifications:
            n = pynotify.Notification(summary, body, icon_name, self)
            n.show()
        return
        
    def check_connection (self):
        try:
            self.__dbus_iface.GetDaemonVersion()
        except:
            print "Daemon doesn't respond to GetDaemonVersion... trying to reconnect"
            # Connection failed, try to reconnect
            self.__cleanup()
            self.__connect()
        
        return
    

    def __calculate_popdown_menu_position (self, menu, data=None):    
        (x, y) = self.get_parent_window().get_origin()
        (menu_w, menu_h) = menu.size_request()
        (applet_w, applet_h) = (self.allocation.width, self.allocation.height)
        orient = self.get_orient()
        
        if orient == gnomeapplet.ORIENT_DOWN:
            y += applet_h
            x -= (menu_w - applet_w) / 2
        elif orient == gnomeapplet.ORIENT_UP:
            y -= menu_h
            x -= (menu_w - applet_w) / 2
        elif orient == gnomeapplet.ORIENT_RIGHT:
            y -= (menu_h - applet_h) / 2
            x += applet_w
        elif orient == gnomeapplet.ORIENT_LEFT:
            y -= (menu_h - applet_h) / 2
            x -= menu_w
            
        return (x, y, False)
    
    def __refresh_icon (self):
        size = self.__panel_size - 4
        
        # Don't let icon size go below 16...
        size = max(size, 16)
        
        scaled_pixbuf = self.__pixbuf_icon.scale_simple(size, size, gtk.gdk.INTERP_TILES)
        
        if self.__connected == False:
            # Desaturate (= greyscale) the image
            scaled_pixbuf.saturate_and_pixelate(scaled_pixbuf, 0.00, False)
                
        self.__image.set_from_pixbuf(scaled_pixbuf)
    
    def __cleanup (self):                
        # Some variables
        self.__connected = False
        
        # Destroy devices
        for device in self.__devices:
            device.destroy()
        self.__devices = []
        
        # Destroy menu
        if self.__menu:
            self.__menu.destroy()
        self.__menu = None
        
        # D-BUS proxy and interface
        self.__dbus_proxy_obj = None
        self.__dbus_iface = None
        
        self.__refresh_icon()
        
        return
    
    def __switch_bus (self):        
        # Close existing D-BUS connection
        if self.__dbus_bus:
            self.__signal_started.remove()
            self.__dbus_bus = None
        
        # Init D-BUS connection
        if self.__use_system_bus == True:
            print "Using system bus"
            self.__dbus_bus = dbus.SystemBus()
        else:
            print "Using session bus"
            self.__dbus_bus = dbus.SessionBus()
            
        # Set up signal handlers
        self.__signal_started = self.__dbus_bus.add_signal_receiver(self.__signal_daemon_started,
            'DaemonStarted',
            'net.sf.cdemu.CDEMUD_Daemon',
            None, # Don't request interface because we don't need it at this point
            '/CDEMUD_Daemon')
        
        return
    
    def __connect (self):        
        # Get D-BUS interface
        try:
            self.__dbus_proxy_obj = self.__dbus_bus.get_object('net.sf.cdemu.CDEMUD_Daemon', '/CDEMUD_Daemon')
            self.__dbus_iface = dbus.Interface(self.__dbus_proxy_obj, 'net.sf.cdemu.CDEMUD_Daemon')
        except dbus.DBusException, e:
            # Display dialog only for errors other than "Service not running"
            if e.get_dbus_name() != "org.freedesktop.DBus.Error.ServiceUnknown":
                self.__display_error(_("Failed to connect: %s") % e.message)
            else:
                print "Service not running"
            
            return False
        
        # Get daemon interface version
        try:
            interface_version = self.__dbus_iface.GetDaemonInterfaceVersion()
        except dbus.DBusException, e:
            self.__display_error(_("Failed to acquire daemon interface version (this most likely means your daemon is out-of-date): %s") % e.message)
            return False
        
        # Check daemon version; we should get away with direct string comparison
        # in Python...
        if (interface_version != config.daemon_interface_version):
            self.__display_error(_("CDEmu daemon interface version %i detected, but version %i is required!") % (interface_version, config.daemon_interface_version))
            return False
        
        # Signal receivers
        self.__dbus_iface.connect_to_signal("DaemonStopped", self.__signal_dbus_daemon_stopped)
                
        # Get number of devices
        try:
            num_devices = self.__dbus_iface.GetNumberOfDevices()
        except dbus.DBusException, e:
            self.__display_error(_("Failed to acquire number of devices: %s") % e.message)
            return False
        
        # Create menu
        self.__menu = gtk.Menu()
        
        for i in range(0, num_devices):
            # Create device
            self.__devices.append(gCDEmu_Device(i, self, self.__menu, self.__dbus_iface))
        
        # We're officially connected now
        self.__connected = True
        
        # Refresh icon
        self.__refresh_icon()
        
        return True

gobject.type_register(gCDEmu_Applet)
